3.x Set default locale via Middleware

iwasherefirst2 opened the pull request:
See https://github.com/mcamara/laravel-localization/issues/921

jordyvanderhaegen reviewed (commented):
Hi @iwasherefirst2, great work so far! 👍 
I've gone ahead and shared my thoughts on the proposed changes.

> Update the entire README

Sounds good, let's delay entire rewrites until 3.x is almost ready though.

> Update tests

I've made some changes myself:
- Macro registering does not throw exceptions whilst running tests anymore. This was caused by the macro trying to register itself multiple times. We now check if it has been registered previously, otherwise we discard it.
- Tests now use testbench's default testcase instead of the browserkit one.
- Tests now register routes using the new `localized` macro.
- Tests now do not refresh the application anymore, as this should not be necessary 
with the new way of registering routes.
- Tests now use `get` instead of `call`, which should reduce the amount of code to get the same result.

Some tests are still failing, but that seems to be related to the changes in this PR and should be addressed.

> Create a Docker image to run the tests locally (would be very helpful)

Is there a reason you would implement docker here? Doesn't seem necessary to me.
We already have tests running through GitHub Actions for any PR's targetting master.

> Perform smoke tests for each feature

Some tests will have to be changed in order to match the new functionality. I wouldn't mind looking into this.

iwasherefirst2 commented:
> * browserkit



> Hi @iwasherefirst2, great work so far! 👍 I've gone ahead and shared my thoughts on the proposed changes.
> 
> > Update the entire README
> 
> Sounds good, let's delay entire rewrites until 3.x is almost ready though.

Sure
> 
> > Update tests
> 
> I've made some changes myself:
> 
>     * Macro registering does not throw exceptions whilst running tests anymore. This was caused by the macro trying to register itself multiple times. We now check if it has been registered previously, otherwise we discard it.
> 
>     * Tests now use testbench's default testcase instead of the browserkit one.
> 
>     * Tests now register routes using the new `localized` macro.
> 
>     * Tests now do not refresh the application anymore, as this should not be necessary
>       with the new way of registering routes.
> 
>     * Tests now use `get` instead of `call`, which should reduce the amount of code to get the same result.
> 

These are all great changes!

> 
> Some tests are still failing, but that seems to be related to the changes in this PR and should be addressed.

I also had to make one change after your commits. I realized that in version 2.x, localized routes without a locale in the URL always worked, even when hideDefaultLocaleInURL was disabled. I'm not sure if this was intended behavior or just a side effect/bug. In addition, I added some validation for the `{locale}` parameter.

In my opinion, routes within the localized group should not be accessible without a locale if hideDefaultLocaleInURL is disabled. Allowing this could result in the language being determined by the browser's language settings  (without redirect), leading to inconsistencies for bots, which SEO algorithms typically penalize. Since we're now in 3.x, it might be a good opportunity to revisit and adjust this behavior. What do you think? For now, I’ve reverted the behavior to match 2.x for consistency, but I don’t think it makes much sense.

> 
> > Create a Docker image to run the tests locally (would be very helpful)
> 
> Is there a reason you would implement docker here? Doesn't seem necessary to me. We already have tests running through GitHub Actions for any PR's targetting master.

My local PHP version doesn’t have all the libraries required to run the Orchestra test suite. Currently, I’ve set up the package inside a fresh Laravel application, so I use Sail to run the tests, and it works fine for me now.

That said, I think having a Docker image could still be helpful for contributors. It would allow them to run tests locally without relying on the GitHub runner, which can save time during development. However, it’s not essential, and if you’d prefer not to use it, that’s fine. If I find the time, I might add a Dockerfile and some instructions to the README for future use.

> 
> > Perform smoke tests for each feature
> 
> Some tests will have to be changed in order to match the new functionality. I wouldn't mind looking into this.

Yes, it looks like it.

iwasherefirst2 commented on .idea/.gitignore:
Absolutely, we should not track them. Was added by mistake.

iwasherefirst2 commented on README.md:
```diff
@@ -8,6 +8,11 @@
 
 Easy i18n localization for Laravel, an useful tool to combine with Laravel localization classes.
 
+This documentation covers version 3.x, compatible exclusively with Laravel 9 and later.
+We strongly recommend using this version, as it resolves numerous compatibility issues present in versions 1.x and 2.x, 
+and seamlessly integrates with Laravel's default caching. 
+If you need to use an older version, please switch to `legacy` branch by clicking [here](https://github.com/mcamara/laravel-localization/tree/legacy).
+
```
Upgrade 3.x introduces significant changes, and my goal is to maintain a clean and concise README for this version. Including setup details for 1.x-2.x alongside 3.x would unnecessarily bloat the README, as the setups differ substantially

My idea was to streamline the README by focusing exclusively on 3.x, while users requiring older versions could switch to the 2.x or legacy branch. These branches would retain the current README from master, similar to how Laravel organizes branches for each major version (e.g., 11.x, 10.x).

Since we're also removing files like caching-related documents in 3.x, I thought this approach would make it more convenient for users sticking with an outdated Laravel and version 2.x.

However, if this doesn’t resonate, I’m open to an alternative: creating a separate `LEGACY_README.md` for installation instructions and other details for 1.x-2.x. My primary objective is to keep the 3.x README clean and focused.

jordyvanderhaegen replied:
@iwasherefirst2 I don't like the idea of having a legacy readme file. It goes against the principles of git , as it's just a copy of an older version. Instead, I've gone ahead and created branches for all major versions, 0.x, 1.x, 2.x, etc and referred to those from the readme itself. This way, we refer to the entire repository and documentation for the specific version.

iwasherefirst2 replied:
I agree, that is the best option. Thanks!

iwasherefirst2 commented on src/Mcamara/LaravelLocalization/LaravelLocalization.php:
```diff
@@ -125,20 +78,13 @@ class LaravelLocalization
-        Router $router,
-        Request $request,
-        UrlGenerator $url
+        private readonly Application $app,
+        private readonly ConfigRepository $configRepository,
+        private readonly Translator $translator,
+        private readonly Router $router,
+        private readonly Request $request,
+        private readonly UrlGenerator $url
```
You’re right—these should be protected for extensibility. I’m leaning toward keeping readonly, as components like Application and Router aren’t typically replaced. However, if needed, you can change it to just protected to allow reassignment.

iwasherefirst2 replied:
updated visibilty to protected

jordyvanderhaegen replied:
@iwasherefirst2 have you pushed these changes? :)

iwasherefirst2 replied:
pushed now directly to 3.x

iwasherefirst2 commented on src/Mcamara/LaravelLocalization/LaravelLocalization.php:
```diff
@@ -585,22 +477,6 @@ public function getCurrentLocale()
         return $this->configRepository->get('app.locale');
     }
 
-    /**
-     * Returns current regional.
-     *
-     * @return string current regional
-     */
-    public function getCurrentLocaleRegional()
-    {
-        // need to check if it exists, since 'regional' has been added
-        // after version 1.0.11 and existing users will not have it
-        if (isset($this->supportedLocales[$this->getCurrentLocale()]['regional'])) {
-            return $this->supportedLocales[$this->getCurrentLocale()]['regional'];
-        } else {
-            return;
-        }
-    }
-
```
True. If its mentioned in the readme, we probably should keep it.

iwasherefirst2 replied:
resolved

iwasherefirst2 commented on src/Mcamara/LaravelLocalization/Middleware/LocaleSetLanguage.php:
```diff
@@ -0,0 +1,109 @@
+        return $next($request);
+    }
+
+    protected function getSupportedLocales()
+    {
+        $locales = $this->configRepository->get('laravellocalization.supportedLocales');
+
+        if (empty($locales) || !\is_array($locales)) {
+            throw new SupportedLocalesNotDefined();
+        }
+
+        return $locales;
+    }
```
We should use it from `LaravelLocalization` class and remove it from the`middleware`. Initially, I thought it could be removed from `LaravelLocalization` later, but that’s not possible since it’s actively used and mentioned in the documentation.

iwasherefirst2 replied:
Restored

iwasherefirst2 commented on src/Mcamara/LaravelLocalization/Middleware/LocaleSetLanguage.php:
```diff
@@ -0,0 +1,109 @@
+    }
+
+
+    protected function hideDefaultLocaleInURL()
+    {
+        return $this->configRepository->get('laravellocalization.hideDefaultLocaleInURL');
+    }
```
you are right

iwasherefirst2 commented on src/Mcamara/LaravelLocalization/Middleware/LocaleSetLanguage.php:
```diff
@@ -0,0 +1,109 @@
+    }
+
+
+    protected function useAcceptLanguageHeader()
+    {
+        return $this->configRepository->get('laravellocalization.useAcceptLanguageHeader');
+    }
```
agree

iwasherefirst2 commented on src/Mcamara/LaravelLocalization/Middleware/LocaleSetLanguage.php:
```diff
@@ -0,0 +1,109 @@
+        return $this->configRepository->get('laravellocalization.useAcceptLanguageHeader');
+    }
+
+    protected function getLocaleRegional(string $locale): string|null
+    {
+        // need to check if it exists, since 'regional' has been added
+        // after version 1.0.11 and existing users will not have it
+        if (!isset($this->getSupportedLocales()[$locale]['regional'])) {
+            return null;
+        }
+
+        return $this->getSupportedLocales()[$locale]['regional'];
+    }
```
ok

iwasherefirst2 commented on src/Mcamara/LaravelLocalization/Middleware/LocaleSetLanguage.php:
```diff
@@ -0,0 +1,109 @@
+use Mcamara\LaravelLocalization\LanguageNegotiator;
+use Mcamara\LaravelLocalization\LaravelLocalization;
+
+class LocaleSetLanguage extends LaravelLocalizationMiddlewareBase
```
`SetLocale` is a much better choice.

iwasherefirst2 commented:
It seems that implementing the [translated routes](https://github.com/mcamara/laravel-localization?tab=readme-ov-file#translated-routes) feature is a bit tricky with the new approach. I have an idea that I'd like to propose in a separate MR.

Would it make sense to merge this for now and continue iterating on separate branches? This way, it’ll be easier to review changes and avoid conflicts between commits.

jordyvanderhaegen commented:
@iwasherefirst2 

> In my opinion, routes within the localized group should not be accessible without a locale if hideDefaultLocaleInURL is disabled. Allowing this could result in the language being determined by the browser's language settings (without redirect), leading to inconsistencies for bots, which SEO algorithms typically penalize. Since we're now in 3.x, it might be a good opportunity to revisit and adjust this behavior. What do you think? For now, I’ve reverted the behavior to match 2.x for consistency, but I don’t think it makes much sense.

I agree with you on this, this doesn't make much sense at first glance. If there's no tests that covered this functionality, let's consider it a bug / unexpected behaviour. Can we make sure to provide tests to ensure this works as expected?

> My local PHP version doesn’t have all the libraries required to run the Orchestra test suite. Currently, I’ve set up the package inside a fresh Laravel application, so I use Sail to run the tests, and it works fine for me now. That said, I think having a Docker image could still be helpful for contributors. It would allow them to run tests locally without relying on the GitHub runner, which can save time during development. However, it’s not essential, and if you’d prefer not to use it, that’s fine. If I find the time, I might add a Dockerfile and some instructions to the README for future use.

I see, that's a bit cumbersome. Feel free to tag me in the PR if you'd like a review.

> Would it make sense to merge this for now and continue iterating on separate branches? This way, it’ll be easier to review changes and avoid conflicts between commits.

Sure, 3.x is very much a WIP.

jordyvanderhaegen commented:
@iwasherefirst2 I'm going to merge this, but won't delete the branch yet. If you have open changes on this branch, you can directly merge them in 3.x.

iwasherefirst2 commented:
Thanks for merging @jordyvanderhaegen , I just added one commit for the visibility of constructor properties.

jordyvanderhaegen commented:
@iwasherefirst2 great, no point in keeping the remote branch then, I deleted it 😄
